cupy/cutensor.py

```diff
@@ -7,27 +7,33 @@
 from cupy.cuda import runtime
 
 _handles = {}
-
-
-def get_handle():
-    dev = device.get_device_id()
-    if dev not in _handles:
-        _handles[dev] = cutensor.create()
-    return _handles[dev]
+_tensor_descriptors = {}
+_contraction_descriptors = {}
+_contraction_finds = {}
+_contraction_plans = {}
```

Comment:
Are they thread safe?

Reply:
@anaruse

> Are they thread safe?

Let me elaborate more on this; it is unclear from cuTENSOR docs whether we can call cuTENSOR routines using the same handle/descriptors/... concurrently from multiple threads. If is unsafe, we should use thread-local storage to avoid sharing these like as done in #1109 (#2053).

Reply:
I'm not if it is thread-safe or not. Please wait for a while.

Reply:
Yes, they do not have to be thread safe during initialiation.

Reply:
Race during initialization can be accepted as long as there's no memory leak (I suppose so as far as I read the current implementation).

Reply:
The handle, descriptor and plan are all thread-safe, in other word, read-only objects after initialization. So, it is safe to share the object among threads (as long as it is not allocated on the stack of a thread).

Reply:
Thank you for clarification!
I found those objects are passed to API functions by const references, so it makes sense.